pkg/sqlcache/informer/listoption_indexer.go

```diff
@@ -323,28 +321,90 @@ func NewListOptionIndexer(ctx context.Context, s Store, opts ListOptionIndexerOp
 func (l *ListOptionIndexer) GetLatestResourceVersion() []string {
 	var latestRV []string
 
-	l.latestRVLock.RLock()
+	l.lock.RLock()
 	latestRV = []string{l.latestRV}
-	l.latestRVLock.RUnlock()
+	l.lock.RUnlock()
 
 	return latestRV
 }
 
+func watcherWithBackfill[T any](ctx context.Context, eventsChan chan<- T) (chan<- T, func()) {
+	backfill, backfillDone := context.WithCancel(ctx)
+	writeChan := make(chan T)
+	buffer := make(chan T, 100)
+
+	// Store any input in the buffer during backfilling.
+	// Immediately stop accepting new events once backfilling completed
+	// Consuming the buffer needs to happen in another goroutine to avoid deadlocks
+	go func() {
+		defer close(buffer)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-backfill.Done():
+				return
+			case event := <-writeChan:
+				buffer <- event
```

Comment:
Just an observation and if you could confirm that'd be great too.

After 100 events, this write will block until we've read from the DB (backfilling done). This will in turn block the send to watchers in `notifyEvent` which will also block ingestion.

I think it's unlikely that we'll receive 100 events before we finish reading from the DB.. What do you think?

Reply:
Correct, and that's the reason why I used a buffered channel and not just a slice, so it could block if the buffer became full, and prevent a possible memory peak (e.g. event storm creating a big buffer slice).